pages/_app/index.jsx

```diff
@@ -78,41 +80,39 @@ class App extends NextApp {
     }
   }
 
-  fetchUser = async () => {
-    try {
-      const { dataProvider, activity } = new Route(window.location.pathname)
-      await this.store.init(dataProvider, activity)
-      this.setState({ isAuthorized: true })
-      Sentry.configureScope((scope) => {
-        scope.setUser({
-          id: this.store.user.id,
-          email: this.store.user.email,
-        })
-      })
-    } catch (unauthorizedError) {
-      // TODO: Do some check before redirect
-      this.setState({ isAuthorized: false })
-      if (!this.isRouteWithoutStore) this.redirectToLogin()
-      throw unauthorizedError
-    }
-  }
-
   get isRouteWithoutStore() {
     const { router } = this.props
     return ROUTES_WITHOUT_STORE.find((r) => router.asPath.startsWith(r))
   }
 
+  get isAuthorized() {
+    return !!this.store?.user?.id
+  }
+
   async componentDidMount() {
     const { router } = this.props
+    router.events.on('routeChangeStart', this.handleRouteChange)
 
     window.addEventListener('unhandledrejection', this.handlePromiseRejection)
 
-    if (!this.isRouteWithoutStore) await this.fetchUser()
+    if (this.isRouteWithoutStore) return
 
-    // we use routeChangeStart to immediately update selected activity
-    // ensure it's bound after user is fetched to avoid false errors
-    //  - user has no access to data provider
-    router.events.on('routeChangeStart', this.handleRouteChange)
+    try {
+      await this.store.init()
+
+      Sentry.configureScope((scope) => {
+        scope.setUser({
+          id: this.store.user.id,
+          email: this.store.user.email,
+        })
+      })
+
+      // Re-renders the component
+      // Initialises router.query with the current URL
+      this.softReload()
```

Comment:
So basically [nextjs keeps query empty in componentDidMount](https://nextjs.org/docs/routing/dynamic-routes#caveats) and fill it in next update. It's really annoying.


> Pages that are statically optimized by Automatic Static Optimization will be hydrated without their route parameters provided, i.e query will be an empty object ({}).
> 
> After hydration, Next.js will trigger an update to your application to provide the route parameters in the query object.

Reply:
Thanks a lot! It's really helpful even if I do not know what to do with this now. I wonder how did you find this?

Reply:
Well we can opt-out from Automatic Static Optimization for dynamic pages, i.e we need to add getInitialProps to _app.jsx or add getInitialProps to all pages in /data-provides/ folder. Eventually which is the best solution IMHO is to parse data provider id manually from windows.location in componentDidMount. 

I found it by browsing issues in next repo.

Reply:
It seems to me that we should do the following:
1. keep `componentDidMount` sync
2. trigger store initialisation sync
3. subscribe for the `user` update (observe it)
4. once the `user` is changed, run route initialisation
   - at this stage, the router should be already dehydrated (🤔❓), `query` object is not empty

In the algorithm, I rely on the fact that there is an extremely small chance to have `user` initialised bebore router's `query` is. However, it's better to somehow observe the router. I wonder if the router triggers some events after initialisation, maybe `routeChangeComplete`.

Reply:
Sounds to me too complicated in comparison with the current state. (just parse provider from window object) Also, it's not good to rely on that fact even when there is a small chance... It can cause really hard debugging bugs.

Reply:
It's actually not. Next does work as I expected, `routeChangeComplete` fires after the client initialisation (dehydration, whatever). I subscribe to 2 promises and then run updates, as simple as it is; [check it out](https://github.com/oacore/dashboard/pull/297/commits/09cdb2ec4da419bbee1e0feba37b22e31a3def47).

Reply:
Interesting. I though routeChangeStart and Complete are called always together. Is it just your assumption or did you find a proof for it? It may work like that just because routeChangeStart was fired before we managed to subscribe for it and since there is a delay between start and complete we manage to subscribe for the complete event. :)

Reply:
I ran a simple test with console logs marking each event. After initialisation only and always complete-one fires.

Reply:
I left `console.log` in my example. Feel free to check it, if everything is fine, I consider this PR ready

Reply:
Ok, you still haven't convinced me but let's see if it brings bugs. There are no much options anyway and this is the cleanest solution. :)